Upgrade to PSR12 coding standard

razvanphp opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | yiisoft/yii2-coding-standards#42

razvanphp commented on framework/web/Request.php:
```diff
@@ -1008,9 +1007,17 @@ private function utf8Encode($s)
+                default: $s[$j] = "\xC3";
+                    $s[++$j] = \chr(\ord($s[$i]) - 64);
+
+break;
```
this indeed looks worse....

razvanphp replied:
Fixed manually.

razvanphp commented:
Funny thing is... because of `PSR12.Files.FileHeader.SpacingAfterBlock` we need to change **every** file to add a new line before the copyright docblock....

samdark commented on framework/console/Controller.php:
```diff
@@ -222,8 +222,7 @@ public function bindActionParams($action, $params)
                 unset($params[$key]);
-            } elseif (
-                PHP_VERSION_ID >= 70100
+            } elseif (PHP_VERSION_ID >= 70100
```
It was better before the change. Is it part of PSR-12?

razvanphp replied:
No, PSR2 only, I did not push anything for PSR12 yet.

razvanphp replied:
I think it checks how the first condition is, see few lines before.

razvanphp replied:
With PSR12 this line is not touched anymore.

samdark commented on framework/console/controllers/MessageController.php:
```diff
@@ -436,38 +436,38 @@ protected function saveMessagesToDb($messages, $db, $sourceMessageTable, $messag
+            foreach ($freshMessagesIds as $id) {
+                if (!in_array($id, $msgRowsIds)) {
+                    $db->createCommand()
+                     ->insert($messageTable, ['id' => $id, 'language' => $language])
```
This is looking worse than it was.

razvanphp replied:
Yeah, we need to tweak the standards first to avoid those useless changes.

razvanphp replied:
This is just increasing the indent, it was wrong before, see this vscode diff:
<img width="1365" alt="Screenshot 2024-03-03 at 11 42 25" src="https://github.com/yiisoft/yii2/assets/4599319/05ce2d4d-3f9e-4e7f-8e93-c9a2bc18f466">

samdark replied:
Nope. See this:

<img width="292" alt="image" src="https://github.com/yiisoft/yii2/assets/47294/dbfb9166-6c93-41fa-ad78-15f4fbacaa2c">

It was properly indented but now it's not.

rob006 replied:
Comments have wrong indention too.

razvanphp replied:
Fixed this one manually, please check again.

samdark commented on framework/console/controllers/MessageController.php:
```diff
@@ -477,15 +477,15 @@ protected function saveMessagesToDb($messages, $db, $sourceMessageTable, $messag
-          if (!in_array($currentLanguage, $languages)) {
-            $deleted=$db->createCommand()->delete($messageTable, "language=:language", [
+            if (!in_array($currentLanguage, $languages)) {
+                $deleted=$db->createCommand()->delete($messageTable, 'language=:language', [
```
No spaces around `=`.

razvanphp replied:
This will be fixed by PSR12, I did not push it yet.

samdark commented on framework/db/BaseActiveRecord.php:
```diff
@@ -311,8 +311,7 @@ public function __get($name)
         if ($this->hasAttribute($name)) {
-            if (
-                !empty($this->_relationsDependencies[$name])
+            if (!empty($this->_relationsDependencies[$name])
```
Got worse.

razvanphp replied:
I first applied PSR2. Now I force pushed with PSR12 and this line is not changed anymore.

samdark commented on framework/web/Request.php:
```diff
@@ -1008,9 +1007,17 @@ private function utf8Encode($s)
-                default: $s[$j] = "\xC3"; $s[++$j] = \chr(\ord($s[$i]) - 64); break;
+                case $s[$i] < "\x80": $s[$j] = $s[$i];
+
+break;
```
:(

razvanphp replied:
Ignored this function in the docblock.

samdark commented on framework/web/Request.php:
```diff
@@ -1008,9 +1007,17 @@ private function utf8Encode($s)
+                    $s[++$j] = $s[$i];
+
+break;
+                default: $s[$j] = "\xC3";
```
Shouldn't it be on the next line?

razvanphp replied:
I did not do any manual changes, only the automatic ones.

razvanphp replied:
we can prevent this for this function if we add those to the docblock:
```
     * @phpcs:disable Generic.Formatting.DisallowMultipleStatements.SameLine
     * @phpcs:disable Squiz.WhiteSpace.ScopeClosingBrace.ContentBefore
```
I guess this is ok, right? it looks ugly if we try to refactor it.

samdark replied:
How about excluding these rules?

rob006 replied:
Did you tried to fix it manually? Maybe fixer is not able to fix it correctly, but it won't mess with code style if there is nothing to fix.

razvanphp replied:
Ignored this function in the docblock.

rob006 commented on framework/db/BaseActiveRecord.php:
```diff
@@ -311,8 +311,7 @@ public function __get($name)
         if ($this->hasAttribute($name)) {
-            if (
-                !empty($this->_relationsDependencies[$name])
+            if (!empty($this->_relationsDependencies[$name])
```
Which rule is responsible for such change? It looks worse and it is inconsistent with similar constructs (like list of arguments in function call).

razvanphp replied:
Ideed. i will revert and check the squizz rules responsible.

razvanphp replied:
This was it:
```
FILE: /Volumes/Code/yii2/framework/db/BaseActiveRecord.php
-------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
-------------------------------------------------------------------------------------------------------------------------------------------------
 314 | ERROR | [x] Expected 0 spaces after opening bracket; newline found (PSR2.ControlStructures.ControlStructureSpacing.SpacingAfterOpenBrace)
 524 | ERROR | [x] Expected 0 spaces after opening bracket; newline found (PSR2.ControlStructures.ControlStructureSpacing.SpacingAfterOpenBrace)
-------------------------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-------------------------------------------------------------------------------------------------------------------------------------------------
```
The good thing is, with PSR12, there is no change, I wanted to do it in separate commits...

rob006 commented on framework/views/_createTable.php:
```diff
@@ -11,9 +11,9 @@
     if (empty($field['decorators'])): ?>
             '<?= $field['property'] ?>',
-<?php else: ?>
+    <?php else: ?>
```
This looks incorrect - it will generate template with incorrect formatting.

razvanphp replied:
I think it will work, as the php code is not in the final template, but still, it’s useless change.

rob006 replied:
I'm pretty sure it will add unnecessary tab, which will mess up with indentation of generated migration. In `framework/views/_foreignTables.php` is similar problem, but it affects phpdoc.

razvanphp replied:
```
FILE: /Volumes/Code/yii2/framework/views/_createTable.php
-----------------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
-----------------------------------------------------------------------------------------------------------------------------
  5 | ERROR | [x] Header blocks must be separated by a single blank line (PSR12.Files.FileHeader.SpacingAfterBlock)
 14 | ERROR | [x] Closing brace indented incorrectly; expected 4 spaces, found 0 (Squiz.WhiteSpace.ScopeClosingBrace.Indent)
-----------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-----------------------------------------------------------------------------------------------------------------------------
```

rob006 replied:
I think that generating correct output is much more important than following some CS rules, so I suggest to disable views from CS validation/fixing.

razvanphp replied:
fixed in [5b9eb7d](https://github.com/yiisoft/yii2/pull/20121/commits/5b9eb7df339296163905bd0c3bf083f9c08ea752)

razvanphp committed:
Add phpcs coding standard deps + github action

razvanphp committed:
Fix PSR12 code style errors

razvanphp committed:
Fix PSR12.Files.FileHeader.SpacingAfterBlock

This touches every file because of the copyright header missing space.

@see PSR12 Chapter 3.

razvanphp commented:
Only `PSR12.Properties.ConstantVisibility.NotFound` left, I will put all of them to public, which is implied/default already.

rob006 commented on framework/helpers/BaseFormatConverter.php:
```diff
@@ -240,63 +241,63 @@ public static function convertDatePhpToIcu($pattern)
             // Day
             '\d' => "'d'",
-            'd' => 'dd',    // Day of the month, 2 digits with leading zeros 	01 to 31
+            'd' => 'dd',    // Day of the month, 2 digits with leading zeros    01 to 31
```
IMO these tabs should be removed or converted to something other than spaces - I don't see the point in converting them to spaces.

razvanphp replied:
They were automatically fixed. Should we use `[` and `]` maybe?

samdark replied:
`—` should be fine.

razvanphp replied:
Fixed in ad37ec45caa1b82705c2064c200348dabf535f7c

samdark commented on framework/grid/Column.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Is the newline required by PSR?

razvanphp replied:
yes, `PSR12.Files.FileHeader.SpacingAfterBlock`

samdark commented:
Overall it looks good except what's mentioned in comments. Let's fix it and merge.

samdark committed:
Merge branch 'master' into upgrade-to-psr12-coding-standard

razvanphp commented:
@rob006 do you think we should remove the `cs` folder and `friendsofphp/php-cs-fixer` dependancy at this point?
ref: https://github.com/razvanphp/yii2/commit/ba0ab403b52124c941dbeb46fbd9efdc12252a5d

rob006 commented:
@razvanphp Yes, it was never finished and AFAIK never worked as part of CI.

razvanphp committed:
Replace tabs with long dash to comply with phpcs rule

Generic.WhiteSpace.DisallowTabIndent.NonIndentTabsUsed

razvanphp committed:
Remove php-cs-fixer (added in ba0ab40)

razvanphp committed:
Add custom phpcs config to exclude some rules

razvanphp commented:
@samdark this is ready for merge FMPOV, but we need to bump the coding standard repo to use PSR12, then all the tests will pass.

BTW: why is [this](https://github.com/yiisoft/yii2/actions/runs/8283932523/job/22668301269?pr=20121#step:8:71) PHP 5.4 test failing now? seem that indeed [array_column function](https://www.php.net/manual/en/function.array-column.php) was added only in PHP 5.5, so that's why it fails in 5.4, but we did not touch that file....

samdark commented:
Done. Still errors.

razvanphp commented:
> Done. Still errors.

but we need a major tag to bump the coding standards in `composer.json` here

samdark commented:
PHP 5.4 error is very weird cause it doesn't error in master branch and function call is there.

samdark committed:
Merge branch 'master' into upgrade-to-psr12-coding-standard

samdark commented:
Tagged 3.0.0.

razvanphp committed:
Fix test in php54 and bump coding standards to PSR12

razvanphp commented:
Good catch @rob006, fixed the test in [b2585a6](https://github.com/yiisoft/yii2/pull/20121/commits/b2585a683b7b1b6d2660380c67c6c308c7fb1c22)

Now we have to decide what we do with: `Squiz.NamingConventions.ValidVariableName.PrivateNoUnderscore`

The framework code initially [decided](https://github.com/yiisoft/yii2-coding-standards/blob/8bc39acaae848aec1ad52b2af4cf380e3f0b104e/Yii2/ruleset.xml#L7C24-L7C67) to use underscore for private vars, even tho PSR2 had this rule (which is now included in PSR12 too). Should we goo all-in without underscores and adhere to PSR12 fully or add underscore for missing ones?

rob006 commented:
AFAIK underscore in Yii is not to indicate visibility, bot to avoid conflicts with virtual attributes provided by getters and setters. So sometimes you use it, sometimes not, I don't think there is a rule smart enough to cover this.

samdark commented:
Underscores are to be kept. This is a technical thing, not just style.

razvanphp committed:
Ignore phpcs NamingConventions.ValidVariableName.PrivateNoUnderscore for db classes

razvanphp committed:
Refactor private vars with underscore

razvanphp commented:
All checks have passed. Please see my last 2 commits.

samdark commented:
Merged. Thank you for working on it.
